Caboodle/FileSystem/FileSystem.ios.cs

```diff
@@ -0,0 +1,39 @@
+﻿using System;
+using System.IO;
+using System.Threading.Tasks;
+using Foundation;
+
+namespace Microsoft.Caboodle
+{
+    public static partial class FileSystem
+    {
+        public static string CacheDirectory
+            => GetDirectory(NSSearchPathDirectory.CachesDirectory);
+
+        public static string AppDataDirectory
+            => GetDirectory(NSSearchPathDirectory.LibraryDirectory);
+
+        public static string UserDataDirectory
+            => GetDirectory(NSSearchPathDirectory.DocumentDirectory);
+
+        public static Task<Stream> OpenAppPackageFileAsync(string filename)
+        {
+            if (filename == null)
+                throw new ArgumentNullException(nameof(filename));
+
+            filename = filename.Replace("\\", "/");
+            var file = Path.Combine(NSBundle.MainBundle.BundlePath, filename);
+            return Task.FromResult((Stream)File.OpenRead(file));
+        }
+
+        static string GetDirectory(NSSearchPathDirectory directory)
+        {
+            var dirs = NSSearchPath.GetDirectories(directory, NSSearchPathDomain.User);
+            if (dirs == null || dirs.Length == 0)
+            {
+                throw new ArgumentException(nameof(directory));
```

Comment:
This isn't right - the accessor properties will throw if there are missing directories. Properties should not throw.

That said, in normal use this won't fail - but you can create some crazy machines where it will.

Reply:
I do: NSSearchPath.GetDirectories(NSSearchPathDirectory.CachesDirectory, NSSearchPathDomain.User)[0];

Reply:
@praeclarum what would you like the property to do in that case? return null?

Reply:
Yeah that's up to whatever contract @jamesmontemagno wants to sign with the users. ;-) Perhaps breaking the properties don't throw rule is fine here since the other implementations don't seem to handle error cases.

Reply:
Presumably returning null would just delay the exception anyway?  If users aren't checking for null, their code is going to throw if they attempt to use it.